scripts/globals/mining.lua

```diff
@@ -1,149 +1,497 @@
--- Get an item
------------------------------------
+    -- if the item was a coloured rock, turn it into the appropriate element
+    local day = VanadielDayElement();
+    if (item == 769) then
+        item = rocks[day+1];
+    end;
 
-function getMiningItem(player,zone)
-    local Rate = math.random();
+    return item;
+end;
 
-    for zon = 1, table.getn(drop), 2 do
-        if (drop[zon] == zone) then
-            for itemlist = 1, table.getn(drop[zon + 1]), 2 do
-                if (Rate <= drop[zon + 1][itemlist + 1]) then
-                    item = drop[zon + 1][itemlist];
-                    break;
+-------------------------------------------------
+-- Moves mining point.
+-------------------------------------------------
+
+function moveMiningPoint(player,npc,zone)
+
+    local positions = shuffleTable(coordinates[zone]);
+    local points = miningpoints[zone];
+    local newIndex = -1;
+
+    if (positions ~= nil) then
+        -- Generate a list of used indexes for all mining points in this zone.
+        for i,p in ipairs(points) do -- loop through all mining points
+            for c, v in ipairs(positions) do -- find index of this specific point
+                local currentnpc = GetNPCByID(p)
+                if (math.floor(v[1]) == math.floor(currentnpc:getXPos()) and
+                    math.floor(v[2]) == math.floor(currentnpc:getYPos()) and
+                    math.floor(v[3]) == math.floor(currentnpc:getZPos())) then
+                        table.remove(positions,c);
                 end
             end
-            break;
         end
-    end
-
-    --------------------
-    -- Determine result if Colored Rock
-    --------------------
 
-    if (item == 769) then
-        day = VanadielDayElement();
-        item = rocks[day + 1];
+        if (#positions > 0) then
+            newIndex = math.random(1,#positions)
+        end
     end
 
-    --------------------
-    -- Determine chance of no item mined
-    -- Default rate is 50%
-    --------------------
+    if (newIndex == -1) then
+        npc:setLocalVar("HIT_COUNT",math.random(4,6));
+        npc:hideNPC(300);
+        --printf("Couldn't find this mining points current index! Hiding for 5 minutes");
+    else
+        local position = positions[newIndex];
+        npc:setLocalVar("HIT_COUNT",math.random(4,6));
+        npc:hideNPC(120);
+        npc:queue(3000,doMove(npc, position[1], position[2], position[3]));
+    end
+end;
 
-    Rate = math.random();
+-------------------------------------------------
+-- Shuffles a tables elements and returns a copy.
+-------------------------------------------------
 
-    if (Rate <= (1 - MINING_RATE)) then
-        item = 0;
+function shuffleTable(tab)
+    local copy = {}
+    for k, v in pairs(tab) do
+        copy[k] = v
     end
 
-    return item;
+    local res = {}
+    while next(copy) do
+        res[#res + 1] = table.remove(copy, math.random(#copy))
+    end
+    return res
 end
```

Comment:
could you move this function to scripts/globals/common.lua
would be useful to have this for other stuff too